cognite/client/_api/raw.py

```diff
@@ -418,16 +423,24 @@ def _list_generator_concurrent(
             for initial in cursors
         ]
 
-        def exhaust(iterator: Iterator, results: deque[RowList], quit_early: threading.Event) -> None:
+        def exhaust(iterator: Iterator) -> None:
             for res in iterator:
                 results.append(res)
                 if quit_early.is_set():
                     return
+                # User code might be processing slower than the rate of row-fetching, and we want this
+                # iteration-based method to have a upper-bounded memory impact, so we keep a max queue
```

Comment:
Is this really upper-bounded? What if the processing time far exceeds `time.sleep(random.uniform(0, partitions))`, then `results` will just keep growing?

What about something like:
```
while True:
    if (len(results) >= partitions):
        # Sleep randomly per thread to avoid sending all new fetch requests at the same time
        time.sleep(random.uniform(0, partitions))
    else:
        try:
            res = next(iterator)
        except StopIteration:
            return
        results.append(res)
    if quit_early.is_set():
        return
```

Reply:
It is inside a while loop so it will just continuously wait until there's room in `results` (see the screenshot if you don't believe me 😜 ). Your code risks being stuck forever in `sleep` if `quit_early` signal is given.

Reply:
But currently it appends to `results` in every iteration of the loop, regardless of whether `len(results) >= partitions` or not.

Reply:
How does my code get stuck?

Reply:
In your screenshot, the chunk-processing time doesn't exceed 2 seconds, so it's fast enough to not cause `results` to grow beyond some threshold.

Reply:
If we ignore the "limit-is-reached-quit-trigger", the code is this:
```py
for res in iterator:
    results.append(res)
    while len(results) >= partitions:
        time.sleep(random.uniform(0, partitions))
```
Imo, it should stay in the while loop until there's room

Reply:
I read it as:
```
for res in iterator: 
    results.append(res) 
    if len(results) >= partitions:  # <--- `if` not `while`
        time.sleep(random.uniform(0, partitions))
```
🤦

Reply:
Thank you for spoon feeding me 🙏